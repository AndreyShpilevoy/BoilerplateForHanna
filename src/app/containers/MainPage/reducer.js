import {all, call, put, take, fork} from 'redux-saga/effects';
import {getUserProfilesArrayApi, getDivisionsArrayApi} from 'api';

const initialState = {
    userProfilesArray: [],
    divisionsArray: [],
    error: {}
};

export const GET_USER_PROFILE_ARRAY = 'GET_USER_PROFILE_ARRAY';
export const getUserProfilesArray = () => ({type: GET_USER_PROFILE_ARRAY});
export const GET_USER_PROFILE_ARRAY_SUCCESS = 'GET_USER_PROFILE_ARRAY_SUCCESS';
export const getUserProfilesArraySuccess = userProfilesArray => ({type: GET_USER_PROFILE_ARRAY_SUCCESS, payload: {userProfilesArray} });
export const GET_USER_PROFILE_ARRAY_FAILED = 'GET_USER_PROFILE_ARRAY_FAILED';
export const getUserProfilesArrayError = error => ({type: GET_USER_PROFILE_ARRAY_FAILED, payload: {error} });

export const GET_DIVISIONS_ARRAY = 'GET_DIVISIONS_ARRAY';
export const getDivisionsArray = () => ({type: GET_DIVISIONS_ARRAY});
export const GET_DIVISIONS_ARRAY_SUCCESS = 'GET_DIVISIONS_ARRAY_SUCCESS';
export const getDivisionsArraySuccess = divisionsArray => ({type: GET_DIVISIONS_ARRAY_SUCCESS, payload: {divisionsArray} });
export const GET_DIVISIONS_ARRAY_FAILED = 'GET_DIVISIONS_ARRAY_FAILED';
export const getDivisionsArrayError = error => ({type: GET_DIVISIONS_ARRAY_FAILED, payload: {error} });

export const mainPageReducer = (state = initialState, {type, payload}) => {
    switch (type) {
        case GET_USER_PROFILE_ARRAY_SUCCESS:
            return {...state, userProfilesArray: payload.userProfilesArray};
        case GET_USER_PROFILE_ARRAY_FAILED:
            return {...state, error: payload.error};
        case GET_DIVISIONS_ARRAY_SUCCESS:
            return {...state, divisionsArray: payload.divisionsArray};
        case GET_DIVISIONS_ARRAY_FAILED:
            return {...state, error: payload.error};
        default:
            break;
    }
    return state;
};


/* eslint-disable func-style, func-names */
export const getUserProfilesArrayNonBlockSaga = function* () {
    const {userProfilesArray, error} = yield call(getUserProfilesArrayApi);

    if (userProfilesArray) {
        yield put(getUserProfilesArraySuccess(userProfilesArray));
    } else {
        yield put(getUserProfilesArrayError(error));
    }
};

export const getUserProfilesArraySaga = function* () {
    for (;;) {
        yield take(GET_USER_PROFILE_ARRAY);
        yield fork(getUserProfilesArrayNonBlockSaga);
    }
};


export const getDivisionsArrayNonBlockSaga = function* () {
    const {divisionsArray, error} = yield call(getDivisionsArrayApi);
    if (divisionsArray) {
        yield put(getDivisionsArraySuccess(divisionsArray));
    } else {
        yield put(getDivisionsArrayError(error));
    }
};

export const getDivisionsArraySaga = function* () {
    for (;;) {
        yield take(GET_DIVISIONS_ARRAY);
        yield fork(getDivisionsArrayNonBlockSaga);
    }
};

export const mainPageSaga = function* () {
    yield all([
        getUserProfilesArraySaga(),
        getDivisionsArraySaga()
    ]);
};
/* eslint-enable func-style, func-names */