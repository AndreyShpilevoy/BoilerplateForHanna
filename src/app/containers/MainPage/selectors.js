import {createSelector} from 'reselect';
import {sortBy, prop} from 'ramda';

export const userProfilesArraySelector = state => state.mainPageReducer.userProfilesArray;
export const divisionsArraySelector = state => state.mainPageReducer.divisionsArray;

export const sortedUserProfilesArraySelector = createSelector(
    userProfilesArraySelector,
    userProfilesArray => sortBy(prop('order'), userProfilesArray)
);