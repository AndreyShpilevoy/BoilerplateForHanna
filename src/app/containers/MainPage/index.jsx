import React, {PureComponent} from 'react';
import {func, arrayOf, shape, number, string} from 'prop-types';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import Logo from 'components/Logo';
import UserProfile from 'components/UserProfile';
import Division from 'components/Division';
import {getUserProfilesArray, getDivisionsArray} from './reducer';
import {sortedUserProfilesArraySelector, divisionsArraySelector} from './selectors';
import styles from './index.css';

export class MainPagePure extends PureComponent {
    static propTypes = {
        getUserProfilesArray: func.isRequired,
        getDivisionsArray: func.isRequired,
        userProfilesArray: arrayOf(
            shape({
                id: number.isRequired,
                name: string.isRequired,
                order: number.isRequired
            })
        ).isRequired,
        divisionsArray: arrayOf(
            shape({
                id: number.isRequired,
                title: string.isRequired,
                HID: number.isRequired
            })
        ).isRequired
    };

    constructor(props) {
        super(props);
        this.state = {
            pathToFile: 'src/containers/MainPage/index.js'
        };
    }

    componentDidMount() {
        this.props.getUserProfilesArray();
        this.props.getDivisionsArray();
    }

    mapUserProfiles = userProfilesArray => userProfilesArray.map(x => <UserProfile key={x.id} userProfile={x} />);
    mapDivisions = divisionsArray => divisionsArray.map(x => <Division key={x.id} division={x} />);

    render() {
        const {pathToFile} = this.state;
        const {userProfilesArray, divisionsArray} = this.props;
        const title = 'Welcome to React';
        return (
            <div className={styles.app}>
                <div className={styles.appHeader}>
                    <Logo className={styles.appLogo} alt='logo' />
                    <h2>
                        {title}
                    </h2>
                </div>
                <p className={styles.appIntro}>
                    {'To get started, edit '}
                    <code>{pathToFile}</code>
                    {' and save to reload.'}
                </p>
                <div>User profiles list:</div>
                {this.mapUserProfiles(userProfilesArray)}
                <div>Divisions list:</div>
                {this.mapDivisions(divisionsArray)}
            </div>
        );
    }
}


const mapStateToProps = state => ({
    userProfilesArray: sortedUserProfilesArraySelector(state),
    divisionsArray: divisionsArraySelector(state)
});

const mapDispatchToProps = dispatch =>
    bindActionCreators({
        getUserProfilesArray,
        getDivisionsArray
    }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(MainPagePure);
