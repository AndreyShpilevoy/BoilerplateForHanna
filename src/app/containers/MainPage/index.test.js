/* eslint-disable no-undef, max-statements, import/first, react/jsx-filename-extension */

import React from 'react';
import {mount, shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import MainPage, {MainPagePure} from './index';

configure({adapter: new Adapter()});

jest.mock('./reducer', () => ({
    getUserProfilesArray: () => ({type: 'MOCK'}),
    getDivisionsArray: () => ({type: 'MOCK'})
}));

jest.mock('react-router-dom/Link');

describe('MainPage', () => {
    const mockStore = configureMockStore();

    it('MainPagePure without data match expected snapshot', () => {
        const props = {
            userProfilesArray: [],
            divisionsArray: [],
            getUserProfilesArray: () => {},
            getDivisionsArray: () => {}
        };
        expect(shallow(<MainPagePure {...props} />)).toMatchSnapshot();
    });

    it('MainPagePure with data match expected snapshot', () => {
        const props = {
            userProfilesArray: [
                {id: 1, name: 'Ben', order: 1},
                {id: 2, name: 'Suzan', order: 2}
            ],
            divisionsArray: [
                {id: 1, title: 'Soda', HID: 1},
                {id: 2, title: 'Water inc.', HID: 3}
            ],
            getUserProfilesArray: () => {},
            getDivisionsArray: () => {}
        };
        expect(shallow(<MainPagePure {...props} />)).toMatchSnapshot();
    });

    it('MainPage without data match expected snapshot', () => {
        const props = {
            store: mockStore({mainPageReducer: {userProfilesArray: [], divisionsArray: [] } })
        };
        expect(mount(<MainPage {...props} />, {lifecycleExperimental: true})).toMatchSnapshot();
    });

    it('MainPage with data match expected snapshot', () => {
        const props = {
            store: mockStore({mainPageReducer: {
                userProfilesArray: [
                    {id: 1, name: 'Ben', order: 1},
                    {id: 2, name: 'Suzan', order: 2}
                ],
                divisionsArray: [
                    {id: 1, title: 'Soda', HID: 1},
                    {id: 2, title: 'Water inc.', HID: 3}
                ] } })
        };
        expect(mount(<MainPage {...props} />, {lifecycleExperimental: true})).toMatchSnapshot();
    });
});
