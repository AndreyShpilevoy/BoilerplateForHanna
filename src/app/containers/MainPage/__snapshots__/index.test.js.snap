// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage MainPage with data match expected snapshot 1`] = `
<Connect(MainPagePure)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MainPagePure
    divisionsArray={
      Array [
        Object {
          "HID": 1,
          "id": 1,
          "title": "Soda",
        },
        Object {
          "HID": 3,
          "id": 2,
          "title": "Water inc.",
        },
      ]
    }
    getDivisionsArray={[Function]}
    getUserProfilesArray={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    userProfilesArray={
      Array [
        Object {
          "id": 1,
          "name": "Ben",
          "order": 1,
        },
        Object {
          "id": 2,
          "name": "Suzan",
          "order": 2,
        },
      ]
    }
  >
    <div
      className={undefined}
    >
      <div
        className={undefined}
      >
        <Logo
          alt="logo"
          className=""
        >
          <Link
            replace={false}
            to="/MainPage"
          />
        </Logo>
        <h2>
          Welcome to React
        </h2>
      </div>
      <p
        className={undefined}
      >
        To get started, edit 
        <code>
          src/containers/MainPage/index.js
        </code>
         and save to reload.
      </p>
      <div>
        User profiles list:
      </div>
      <UserProfile
        className=""
        key="1"
        userProfile={
          Object {
            "id": 1,
            "name": "Ben",
            "order": 1,
          }
        }
      >
        <div
          className=""
        >
          User id: 1, User name: Ben
        </div>
      </UserProfile>
      <UserProfile
        className=""
        key="2"
        userProfile={
          Object {
            "id": 2,
            "name": "Suzan",
            "order": 2,
          }
        }
      >
        <div
          className=""
        >
          User id: 2, User name: Suzan
        </div>
      </UserProfile>
      <div>
        Divisions list:
      </div>
      <Division
        className=""
        division={
          Object {
            "HID": 1,
            "id": 1,
            "title": "Soda",
          }
        }
        key="1"
      >
        <div
          className=""
        >
          Division id: 1, title: Soda, HID: 1
        </div>
      </Division>
      <Division
        className=""
        division={
          Object {
            "HID": 3,
            "id": 2,
            "title": "Water inc.",
          }
        }
        key="2"
      >
        <div
          className=""
        >
          Division id: 2, title: Water inc., HID: 3
        </div>
      </Division>
    </div>
  </MainPagePure>
</Connect(MainPagePure)>
`;

exports[`MainPage MainPage without data match expected snapshot 1`] = `
<Connect(MainPagePure)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MainPagePure
    divisionsArray={Array []}
    getDivisionsArray={[Function]}
    getUserProfilesArray={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    userProfilesArray={Array []}
  >
    <div
      className={undefined}
    >
      <div
        className={undefined}
      >
        <Logo
          alt="logo"
          className=""
        >
          <Link
            replace={false}
            to="/MainPage"
          />
        </Logo>
        <h2>
          Welcome to React
        </h2>
      </div>
      <p
        className={undefined}
      >
        To get started, edit 
        <code>
          src/containers/MainPage/index.js
        </code>
         and save to reload.
      </p>
      <div>
        User profiles list:
      </div>
      <div>
        Divisions list:
      </div>
    </div>
  </MainPagePure>
</Connect(MainPagePure)>
`;

exports[`MainPage MainPagePure with data match expected snapshot 1`] = `
<div>
  <div>
    <Logo
      alt="logo"
      className=""
    />
    <h2>
      Welcome to React
    </h2>
  </div>
  <p>
    To get started, edit 
    <code>
      src/containers/MainPage/index.js
    </code>
     and save to reload.
  </p>
  <div>
    User profiles list:
  </div>
  <UserProfile
    className=""
    key="1"
    userProfile={
      Object {
        "id": 1,
        "name": "Ben",
        "order": 1,
      }
    }
  />
  <UserProfile
    className=""
    key="2"
    userProfile={
      Object {
        "id": 2,
        "name": "Suzan",
        "order": 2,
      }
    }
  />
  <div>
    Divisions list:
  </div>
  <Division
    className=""
    division={
      Object {
        "HID": 1,
        "id": 1,
        "title": "Soda",
      }
    }
    key="1"
  />
  <Division
    className=""
    division={
      Object {
        "HID": 3,
        "id": 2,
        "title": "Water inc.",
      }
    }
    key="2"
  />
</div>
`;

exports[`MainPage MainPagePure without data match expected snapshot 1`] = `
<div>
  <div>
    <Logo
      alt="logo"
      className=""
    />
    <h2>
      Welcome to React
    </h2>
  </div>
  <p>
    To get started, edit 
    <code>
      src/containers/MainPage/index.js
    </code>
     and save to reload.
  </p>
  <div>
    User profiles list:
  </div>
  <div>
    Divisions list:
  </div>
</div>
`;
